src/Nethermind/Nethermind.Baseline/BaselineModule.cs

```diff
@@ -526,26 +480,27 @@ private BaselineTree RebuildEntireTree(Address treeAddress, Keccak blockHash)
             var insertLeafLogs = _logFinder.FindLogs(insertLeafFilter);
             BaselineTree baselineTree = new ShaBaselineTree(new MemDb(), Array.Empty<byte>(), 5);
 
-            // Keccak leafTopic = new Keccak("0x8ec50f97970775682a68d3c6f9caedf60fd82448ea40706b8b65d6c03648b922");
+            using var batch = baselineTree.StartBatch();
             foreach (FilterLog filterLog in insertLeavesLogs
                 .Union(insertLeafLogs)
                 .OrderBy(fl => fl.BlockNumber).ThenBy(fl => fl.LogIndex))
             {
                 if (filterLog.Data.Length == 96)
                 {
                     Keccak leafHash = new Keccak(filterLog.Data.Slice(32, 32).ToArray());
-                    baselineTree.Insert(leafHash);
+                    baselineTree.Insert(leafHash, false);
```

Comment:
Lukasz suggested inserts on batch -> this way you could avoid loading the data from the DB when recalculating hashes
You currently do

Insert0, CommitDb, Insert1, CommitDb, Insert2, CommitDb, Insert3, CommitDb
Load0, Load1, Insert01, CommitDB, Load2, Load3, Insert23 CommitDb
Load01, Load23, Insert0123, CommitDb

What you could do instead would be
Insert0, Insert1, Insert2, Insert3
Insert01, Insert23
Insert0123,
CommitDb